_posts/2021-10-01-gs_00004_basic_shell_script.md

```diff
@@ -0,0 +1,318 @@
+
+    # Create the project folder
+    mkdir my_project
+
+    # move into the project folder
+    cd my_project
+
+    # create other folders and files
+    mkdir sources
+    mkdir documentation
+    touch README.md
+    touch LICENSE
+
+<div class='note-normal' markdown="1">
+_Note: '.md' is a Markdown extension, which is typical for a README. We will talk about it in a future post._
+</div>
+
+Save the file, add the correct privileges and run it:
+
+    $ chmod +x project_template.sh
+    $ ./project_template.sh
+
+If you list your directory contents, you should see the created project:
+
+    $ ls
+
+So far so good! However, not all projects are called 'my_project' and it would
+be inconvenient to always change the script. That's why we want to introduce you
+to something called **arguments**. If you call your script you can give it
+additional parameters that are used by the script. In particular, we want to
+call our script like this:
+
+    $ ./project_template.sh project_name
+
+Before we can do it, we have to make a tiny change. Within the script, such
+arguments can be accessed with a
+<span>$</span>-sign and a number. <span>$</span>1 would be the first argument,
+<span>$</span>2 the second, and so on. In our script, we need to change the
+following lines:
+
+    # Create the project folder
+    mkdir $1
+
+    # move into the project folder
+    cd $1
+
+Easy, isn't it? Now try to call the script
+
+    $ ./project_template.sh project_name
+
+<div class='note-exercise' markdown="1">
+Exercise:
+
+<span>$</span>0 holds the name of the script. Add a line to the script printing
+the script name using echo and <span>$</span>0.
+Run the script using:
+
+    $ ./project_template.sh hello
+
+</div>
+
+<div class='note-normal' markdown="1">
+_Note: We always use different project names whenever we run the script. What should happen if you create a project that already exists? Probably, the user should get a warning and the creation should be aborted. We will avoid this situation for now and cover such a check in a future post. Also, we should check that the user really entered an argument. Calling the script with no or more than one argument should fail as well._
+</div>
+
+There is one last, but very important aspect. Inside the script, we used a 'cd'
+command, so we should be inside the project folder. However, as you used 'ls'
+before, you might have noticed that this is not the case. You are still where
+you were before you started the script. The reason is the following. If you run
+the script it is not run in your current shell. Internally, a new shell is
+created which is closed at the end of the script. So, the 'cd' has no impact on
+your current shell. Sometimes, however, you want to change it. For your project
+template, it might be useful to directly move into the project folder. If you
+want to do so, you have to **source** the script:
+
+    $ source ./project_template.sh i_am_a_project
+
```

Comment:
Maybe explain what `source` does: the current shell itself will be used to execute the script.